sql/core/src/main/scala/org/apache/spark/sql/diagnostic/DiagnosticListener.scala

```diff
@@ -0,0 +1,103 @@
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.internal.StaticSQLConf.UI_RETAINED_EXECUTIONS
+import org.apache.spark.status.{ElementTrackingStore, KVUtils}
+
+class DiagnosticListener(
+    conf: SparkConf,
+    kvStore: ElementTrackingStore) extends SparkListener {
+
+  kvStore.addTrigger(
+    classOf[ExecutionDiagnosticData],
+    conf.get(UI_RETAINED_EXECUTIONS)) { count =>
+    cleanupExecutions(count)
+  }
+
+  override def onOtherEvent(event: SparkListenerEvent): Unit = event match {
+    case e: SparkListenerSQLExecutionStart => onExecutionStart(e)
+    case e: SparkListenerSQLExecutionEnd => onExecutionEnd(e)
+    case e: SparkListenerSQLAdaptiveExecutionUpdate => onAdaptiveExecutionUpdate(e)
+    case _ => // Ignore
+  }
+
+  private def onAdaptiveExecutionUpdate(event: SparkListenerSQLAdaptiveExecutionUpdate): Unit = {
+    val data = new AdaptiveExecutionUpdate(
+      event.executionId,
+      System.currentTimeMillis(),
+      event.physicalPlanDescription
+    )
+    kvStore.write(data)
+  }
+
+  private def onExecutionStart(event: SparkListenerSQLExecutionStart): Unit = {
+    val planDescriptionMode = ExplainMode.fromString(SQLConf.get.uiExplainMode)
+    val physicalPlan = event.qe.explainString(planDescriptionMode, Int.MaxValue)
+    val data = new ExecutionDiagnosticData(
+      event.executionId,
+      physicalPlan,
+      event.time,
+      None,
+      None
+    )
+    // Check triggers since it's adding new netries
+    kvStore.write(data, checkTriggers = true)
+  }
+
+  private def onExecutionEnd(event: SparkListenerSQLExecutionEnd): Unit = {
+    try {
+      val existing = kvStore.read(classOf[ExecutionDiagnosticData], event.executionId)
+      val planDescriptionMode = ExplainMode.fromString(SQLConf.get.uiExplainMode)
+      val physicalPlan = event.qe.explainString(planDescriptionMode, Int.MaxValue)
+      val data = new ExecutionDiagnosticData(
+        event.executionId,
+        physicalPlan,
+        existing.submissionTime,
+        Some(event.time),
+        event.executionFailure.map(
+          e => s"${e.getClass.getCanonicalName}: ${e.getMessage}").orElse(Some(""))
+      )
+      kvStore.write(data)
+    } catch {
+      case _: NoSuchElementException =>
+      // this is possibly caused by the query failed before execution.
+    }
+  }
+
+  private def cleanupExecutions(count: Long): Unit = {
+    val countToDelete = count - conf.get(UI_RETAINED_EXECUTIONS)
+    if (countToDelete <= 0) {
+      return
+    }
+    val view = kvStore.view(classOf[ExecutionDiagnosticData]).index("completionTime").first(0L)
+    val toDelete = KVUtils.viewToSeq(view, countToDelete.toInt)(_.completionTime.isDefined)
+    toDelete.foreach(e => kvStore.delete(classOf[ExecutionDiagnosticData], e.executionId))
+    kvStore.removeAllByIndexValues(
+      classOf[AdaptiveExecutionUpdate], "id", toDelete.map(_.executionId))
+  }
+}
+
+object DiagnosticListener {
+  val QUEUE_NAME = "diagnostics"
```

Comment:
This is a dedicated queue for this listener?

Reply:
Yes, because collecting/saving the diagnostics could be slow, I don't want it to impact other critical listeners, e.g. UI listener

Reply:
rename to `sqlDiagnostics`

Reply:
@mridulm 
I put this listener in the SQL folder because I need to capture some SQL events. But in the future, this listener can capture other events to provide diagnostics for other components (e.g. executor). So I think a general name may be better.
But `sqlDiagnostics` is also fine to me.